File: Algorithm/Sort/Merge Sort.cpp
#include<iostream>
using namespace std;

void merge(int a[], int left , int mid ,int right ){
    
    int first_arr[mid-left+1];
    int second_arr[right-mid];
    
    for(int i=left,j=0;i<=mid;i++,j++)
        first_arr[j]=a[i];
    
    for(int i=mid+1,j=0;i<=right;i++,j++)
        second_arr[j]=a[i];
    
    int first=0,second=0;
    int arr[right-left+1],index=0;
    
    while(first<(mid-left+1) && second<(right-mid) ){
        if(first_arr[first]<second_arr[second])
            arr[index++]=first_arr[first++];
        else
            arr[index++]=second_arr[second++];
    }
    
    while(first<(mid-left+1))
            arr[index++]=first_arr[first++];
    
    while(second<(right-mid))
            arr[index++]=second_arr[second++];

    for(int x=left,y=0;x<=right;x++,y++)
        a[x]=arr[y];
        
}

void mergesort(int a[],int left,int right){
    if(left>=right)
        return;
    int mid=(left+right)/2;

    mergesort(a,left,mid);
    mergesort(a,mid+1,right);

    merge(a,left,mid,right);
}

int main(){

    int size;
    cin>>size;
    int arr[size];
    for(int& x: arr) cin>>x;

    mergesort(arr,0,size-1);

    for(int x:arr) cout<<x<<" ";
}

File: Algorithm/Sorting/Merge Sort.py
def merge(a,left,mid,right):
    first_arr=[]
    second_arr=[]
    size1=mid-left+1
    size2=right-mid
    
    for i in range(left,mid+1):
        first_arr.append(a[i])
    for i in range(mid+1,right+1):
        second_arr.append(a[i])
    
    result=[]
    index1=0
    index2=0
    
    while(index1<size1 and index2< size2):
        if(first_arr[index1]<second_arr[index2]):
            result.append(first_arr[index1])
            index1+=1
        else:
            result.append(second_arr[index2])
            index2+=1
    
    
    while(index1<size1):
            result.append(first_arr[index1])
            index1+=1
    while(index2<size2):
            result.append(second_arr[index2])
            index2+=1
    
    ind=0
    for i in range(left,right+1):
        a[i]=result[ind]
        ind+=1



def mergeSort(arr,left,right):
    if left >= right:
        return 
    mid=(left+right)//2
    mergeSort(arr,left,mid)
    mergeSort(arr,mid+1,right)
    merge(arr,left,mid,right)

size=int(input(""))
items=list(map(int,input("").split()))

mergeSort(items,0,size-1)


for item in items:
    print(item, end=" ")
    